src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -513,226 +480,118 @@ public RubyRootNode getRootNode() {
-            return deferredCall.call(callNode);
+            return eval(self, rootNodeToEval, cachedCallTarget, callNode, parentFrame);
         }
 
-        @Specialization(guards = {
-                "isRubyString(source)",
-                "isRubyBinding(binding)",
-                "isNil(noFile)",
-                "isNil(noLine)"
-        })
-        public Object evalBinding(VirtualFrame frame, DynamicObject source, DynamicObject binding, DynamicObject noFile, DynamicObject noLine,
-                @Cached("create()") IndirectCallNode callNode) {
-            return evalBinding(frame, source, binding, NotProvided.INSTANCE, NotProvided.INSTANCE, callNode);
-        }
-
-        @Specialization(guards = {
-                "isRubyString(source)",
-                "isRubyBinding(binding)",
-                "isRubyString(file)" })
-        public Object evalBindingFilename(VirtualFrame frame, DynamicObject source, DynamicObject binding, DynamicObject file, NotProvided line,
-                @Cached("create()") IndirectCallNode callNode) {
-            return evalBindingFilenameLine(frame, source, binding, file, 0, callNode);
-        }
+        private Object eval(Object self, RootNodeWrapper rootNode, CallTarget callTarget, DirectCallNode callNode, MaterializedFrame parentFrame) {
+            final InternalMethod method = new InternalMethod(
+                    getContext(),
+                    rootNode.getRootNode().getSharedMethodInfo(),
+                    RubyArguments.getMethod(parentFrame).getLexicalScope(),
+                    rootNode.getRootNode().getSharedMethodInfo().getName(),
+                    RubyArguments.getMethod(parentFrame).getDeclaringModule(),
+                    Visibility.PUBLIC,
+                    callTarget);
 
-        @Specialization(guards = {
-                "isRubyString(source)",
-                "isRubyBinding(binding)",
-                "isRubyString(file)",
-                "isNil(noLine)"
-        })
-        public Object evalBindingFilename(VirtualFrame frame, DynamicObject source, DynamicObject binding, DynamicObject file, DynamicObject noLine,
-                @Cached("create()") IndirectCallNode callNode) {
-            return evalBindingFilename(frame, source, binding, file, NotProvided.INSTANCE, callNode);
+            return callNode.call(RubyArguments.pack(parentFrame, null, method, RubyArguments.getDeclarationContext(parentFrame), null, self, null, new Object[]{}));
         }
 
-        @Specialization(guards = {
-                "isRubyString(source)",
-                "isRubyBinding(binding)",
-                "isRubyString(file)" })
-        public Object evalBindingFilenameLine(VirtualFrame frame, DynamicObject source, DynamicObject binding, DynamicObject file, int line,
+        @Specialization
+        public Object evalBinding(VirtualFrame frame, Object self, DynamicObject source, DynamicObject binding, DynamicObject file, int line,
```

Comment:
The frame does not seem needed in this specialization.
Could you rename it to `evalBindingUncached` and move right below the two other specializations?